### PR TITLE
Include Git hash in the engine

### PR DESCRIPTION
**Pull request by ghost**
I'm not really sure about this PR whether the pros outweigh the cons. It might make issue tracking/regression pinpointing a bit easier? It's intended for those with precompiled binary so that they know which commit exactly they're using. The downside is that compiling takes longer.

screenshot from 2017-07-10 19-10-00

### CODE REVIEW

**Comment by akien-mga** [comment]
> The downside is that compiling takes longer.

Probably *much* longer, as it's changing a header that is included in some core headers, so it would basically recompile all the engine apart from the thirdparty stuff. I think @bojidar-bg had the idea to make it a .cpp file to avoid this issue, though not sure 100% how it would be like.

File: methods.py [hunk 1]
```diff
@@ -1176,6 +1176,12 @@ def update_version():
     f.write("#define VERSION_STATUS " + str(version.status) + "\n")
     import datetime
     f.write("#define VERSION_YEAR " + str(datetime.datetime.now().year) + "\n")
+    head = open(".git/HEAD", "rb").readline().strip()
+    if head.startswith("ref: "):
+        ref = open(".git/" + head[5:], "rb").readline().strip()
+    else:
+        ref = head
```
Comment (akien-mga) [inline methods.py]: That's a bold assumption :) If building from a static archive without .git folder, nothing tells you that it's HEAD (also "head" in itself doesn't mean much, it's <branchname>/HEAD which points to the last commit of the branch). So a better default would be "unknown".

Still, as a distro packager, I can tell you I hate projects which do that :)
If there's no .git folder, just skip this logic and don't display the hash, that's much better.

File: editor/project_manager.cpp [hunk 2]
```diff
@@ -1244,7 +1244,7 @@ ProjectManager::ProjectManager() {
 	top_hb->add_child(ccl);
 	top_hb->add_spacer();
 	l = memnew(Label);
-	l->set_text("v" VERSION_MKSTRING);
+	l->set_text("v" VERSION_MKSTRING "-" + String(VERSION_HASH).left(7));
```
Comment (akien-mga) [inline editor/project_manager.cpp]: You'd also have to add this to `Engine.get_version()`.

**Review by akien-mga — CHANGES_REQUESTED** [review]

File: editor/project_manager.cpp (hunk 2 above)
Comment (akien-mga) [inline editor/project_manager.cpp]: Since the string uses `.` as delimiter between all fields, I guess it might be more consistent to do the same here.

**Comment by Noshyaar** [comment]
Thanks, @akien-mga. :)

Still a bit confused if the default value of `Engine.get_version()` should be `""` or `"unknown"` (or you meant something else from that comment.)

**Comment by akien-mga** [comment]
Well I meant add it to the existing `Engine.get_version_info()` API: https://github.com/godotengine/godot/blob/master/core/engine.cpp#L76

**Comment by Noshyaar** [comment]
dict["string"] there doesn't have the same format as the one on the Project Manager.

File: methods.py (hunk 1 above)
Comment (akien-mga) [inline methods.py]: BTW I misread the logic, I thought you were assigning "head" as string value :)

Comment (Noshyaar) [inline methods.py]: I really should name my variables more descriptively :P

File: methods.py [hunk 3]
```diff
@@ -1177,6 +1177,15 @@ def update_version():
...
     f.write("#define VERSION_YEAR " + str(datetime.datetime.now().year) + "\n")
 
+    ref = ""
+    if os.path.isfile(".git/HEAD"):
+        head = open(".git/HEAD", "rb").readline().strip()
+        if head.startswith("ref: "):
+            ref = open(".git/" + head[5:], "rb").readline().strip()
+        else:
+            ref = head
+    f.write("#define VERSION_HASH \"" + ref + "\"\n")
```
Comment (akien-mga) [inline methods.py]: BTW, how come we don't close the file? :D

**Comment by akien-mga** [comment]
> dict["string"] there doesn't have the same format as the one on the Project Manager.

Yeah, `dict["string"]` is meant for a "pretty" output if you want to put it in your game's logs, etc. But maybe the string of the project manager should be changed to match it, I have no strong opinion.

**Review by akien-mga — DISMISSED** [review]

**Pull request merged**